editor/engine_update_dialog.cpp

```diff
@@ -0,0 +1,219 @@
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "engine_update_dialog.h"
+
+#include "core/os/time.h"
+#include "editor/editor_scale.h"
+#include "editor/editor_settings.h"
+#include "scene/gui/box_container.h"
+#include "scene/gui/button.h"
+#include "scene/gui/label.h"
+#include "scene/main/http_request.h"
+
+void EngineUpdateDialog::_http_request_completed(int p_result, int p_response_code, const PackedStringArray &p_headers, const PackedByteArray &p_body) {
+	if (p_result != OK) {
+		WARN_PRINT(vformat("Failed to check for updates. Error: %s.", p_result));
+		return;
+	}
+
+	if (p_response_code != 200) {
+		WARN_PRINT(vformat("Failed to check for updates. Response code: %s.", p_response_code));
+		return;
+	}
+
+	const Dictionary version_info = Engine::get_singleton()->get_version_info();
+	const String current_version_string = vformat("%s.%s.%s", version_info["major"], version_info["minor"], version_info["patch"]);
+	int current_version = current_version_string.replace(".", "").to_int();
+
+	String max_version_string;
+	int max_version = current_version;
+
+	PackedStringArray lines;
+	{
+		String s;
+		const uint8_t *r = p_body.ptr();
+		s.parse_utf8((const char *)r, p_body.size());
+		lines = s.split("\n");
+	}
+
+	if (!found_version.is_empty()) {
+		update_current_version->set_text(vformat(TTR("Current version: %s"), current_version_string));
+		_check_if_stable(lines);
+		return;
+	}
+
+	for (const String &line : lines) {
+		int i = line.find("href=") + 6;
+		int j = line.find("/\"", i);
+		const String version_string = line.substr(i, j - i);
+
+		int v = version_string.replace(".", "").to_int();
+		if (v < 100) {
+			v *= 10;
+		}
+
+		if (v < 1000 && v > max_version) {
+			max_version = v;
+			max_version_string = version_string;
+		}
+	}
+
+	if (max_version > current_version) {
+		if (max_version_string == EditorSettings::get_singleton()->get_meta("auto_updates_skip_version", "").operator String()) {
+			return;
+		}
+		found_version = max_version_string;
+
+		http->request(vformat("https://downloads.tuxfamily.org/godotengine/%s", found_version));
+	}
+}
+
+void EngineUpdateDialog::_check_if_stable(const PackedStringArray &p_content) {
+	bool is_stable = false;
+	for (const String &line : p_content) {
+		print_line(line);
+		if (line.contains("stable")) {
```

Comment:
If I am currently on a pre-release version I probably do want to update to a newer pre-release, not just the stable version.